src/main/java/graphql/nadel/engine/OverallQueryTransformer.java

```diff
@@ -443,10 +449,50 @@ public TraversalControl enter(TraverserContext<Node> context) {
                 rootVars
         );
 
+        if (removedFieldMap.hasRemovedFields()) {
+            newNode = addUnderscoreTypeNameToEmptySelectionSets(nadelContext, overallTypeInformation, rootVars, treeTransformer, newNode);
+        }
         //noinspection unchecked
         return (T) newNode;
     }
 
+    private Node addUnderscoreTypeNameToEmptySelectionSets(NadelContext nadelContext,
+                                                           OverallTypeInformation<?> overallTypeInformation,
+                                                           Map<Class<?>, Object> rootVars,
+                                                           TreeTransformer<Node> treeTransformer,
+                                                           Node field
+    ) {
+        if (field == null) {
+            return null;
+        }
+        NodeVisitorStub addDummyFieldToEmptySubSelectsVisitor = new NodeVisitorStub() {
+
+            @Override
+            public TraversalControl visitField(Field node, TraverserContext<Node> context) {
+                OverallTypeInfo overallTypeInfo = overallTypeInformation.getOverallTypeInfo(getId(node));
+                if (overallTypeInfo == null) {
+                    return TraversalControl.CONTINUE;
+                }
+                GraphQLFieldDefinition fieldDefinitionOverall = overallTypeInfo.getFieldDefinition();
+                GraphQLNamedOutputType fieldTypeOverall = (GraphQLNamedOutputType) GraphQLTypeUtil.unwrapAll(fieldDefinitionOverall.getType());
```

Comment:
I am wondering if we can make this a bit more clear: do we really need to call the general `maybeAddUnderscoreTypeName`. Can make it more explicit and make a `addEmptySelectionUnderscoeSetTypeName`?

Reply:
Let me try to do that. I am not sure that this is the only place where the typename alias is added to empty selection sets.

Reply:
sorry, might have understood wrong. Do you also want me to get rid of the code that deals with empty selection sets inside of `maybeAddUnderscoreTypeName`?

Reply:
yes if possible, because we don't need it in the general case. Or am I missing something?

Reply:
I think we need it for the removed hydrated fields. When I remove code that deals with empty selection sets from `maybeAddUnderscoreTypeName` two tests are failing in `RemovedFieldsTest.groovy`: 

- all fields are removed from hydrated field
- nested hydrated field is removed

Reply:
for the removed hydrated fields we rely on the fact that `maybeAddUnderscoreTypeName` will add underscore typename to the empty selection sets here:
https://github.com/graphql-java/nadel/blob/master/src/main/java/graphql/nadel/engine/OverallQueryTransformer.java#L126

So we can get rid of the code that deals with empty selection sets inside of `maybeAddUnderscoreTypeName` if we do something like that:

```java
        Field transformedRootField = topLevelField.transform(builder -> builder.selectionSet(topLevelFieldSelectionSet));

        transformedRootField = ArtificialFieldUtils.maybeAddUnderscoreTypeName(nadelContext, transformedRootField, topLevelFieldTypeOverall);
        transformedRootField = ArtificialFieldUtils.maybeAddEmptySelectionSetUnderscoreTypeName(nadelContext, transformedRootField, topLevelFieldTypeOverall);
```

happy to do it if you are ok with it

Reply:
so yeah, I have done it in the latest commit. https://github.com/graphql-java/nadel/pull/171/files#diff-be7fdedc11cd9c188e81f152e70f3e228c48101c61d1478c76a1ec3835d06ef2R133